### PR TITLE
Inject binary into dockerfile

### PR DESCRIPTION
**Pull request by magiconair**
* allow env vars for `archive.name_template`
* inject binary filename into `Dockerfile`

Let me provide some background for this PR and if you agree with the approach then I'll add tests and update the documentation.

I want to use goreleaser to release https://github.com/fabiolb/fabio. I have some special requirements for naming and packaging most of which were already supported by goreleaser. At this point I don't want to change the naming of the artifacts since this would break the debian/ubuntu packages that people are building and which they have automated. 

For fabio I need to publish releases with the binary artifacts for multiple platforms with the following names:

```
fabio-VERSION-go1.x.x-GOOS_GOARCH
fabio-1.5.3-go1.9.2-linux_amd64
```

I also need to publish Docker containers which contain the go version in the tag since I sometimes publish the current version built with a newer go version, e.g.

```
fabiolb/fabio:1.5.3-go1.8.5
fabiolb/fabio:1.5.3-go1.9.2
```

#437 added support for the env vars in the Docker tags which allows me to inject the go version into the Docker tags.

This change now adds support for env vars in the `name_template` so that I can build artifacts with the go version. 

```yaml
archive:
  name_template: '{{ .ProjectName }}-{{ .Version }}-{{ .Env.GOVERSION }}-{{ .Os }}_{{ .Arch }}'
  format: binary
```

However, that breaks the generation of the Docker container since the name of the binary is no longer known upfront in a static Dockerfile. `COPY fabio /` no longer works since the binary is now `fabio-1.5.3-go1.9.2-linux_amd64`. 

The Dockerfile has an `ENV` directive which allows for some templating but I couldn't figure out how to overwrite the ENV variable from the command line with `docker build`. If that is possible then the change becomes much simpler and works outside of goreleaser as well and I would actually prefer that. (https://docs.docker.com/engine/reference/builder/#environment-replacement)

The approach in this PR is to copy the Dockerfile and expand the `${BINARY}` variable in the `Dockerfile` to the true filename of the artifact. Then you can write something like this:

```
FROM scratch
ENV BINARY fabio
COPY ${BINARY} /fabio
ENTRYPOINT ["/fabio"]
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING** document.
- [ ] `make ci` passes on my machine.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 78da472a4594389fdda74792c9b5b9e47409d7a0
  * Your commit message does not contain a body

 * Commit: 3f0e06b1fdd04bfaf26317e9ae2a0b333622a655
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 78da472a4594389fdda74792c9b5b9e47409d7a0
  * Your commit message does not contain a body

 * Commit: 3f0e06b1fdd04bfaf26317e9ae2a0b333622a655
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by caarlos0** [comment]
both changes seem ok to me :)

can you add tests and update the documentation as well?

thanks

**Comment by magiconair** [comment]
Thx and I’ll do that later tonight.

**Comment by magiconair** [comment]
@caarlos0 ready for review.

File: pipeline/docker/testdata/Dockerfile [hunk 1]
```diff
@@ -1,2 +1,2 @@
 FROM scratch
-ADD mybin /
+ADD ${BINARY} /
```
Comment (caarlos0) [inline pipeline/docker/testdata/Dockerfile]: shouldn't this one still be `mybin`?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a1213f98b462df1414726c3f1d426ef8328cb2c3
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a1213f98b462df1414726c3f1d426ef8328cb2c3
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by magiconair** [comment]
yes. changed the wrong file. Pushed a fix.

BTW, the requirement to add a body for every commit runs a bit counter to having small commits. Usually the subject is enough. What's your take on that?

**Comment by caarlos0** [comment]
> BTW, the requirement to add a body for every commit runs a bit counter to having small commits. Usually the subject is enough. What's your take on that?

yeah, I get mad at it sometimes too...
check #446 and #377 for ref

**Comment by caarlos0** [comment]
@magiconair after investigating a bit further, I think that the problem is actually the `archive.format: binary`, not the name template itself...

the binary format + docker integration is broken :(

the name template with environment variables seem good though, I cherry-picked it to #451

**Comment by caarlos0** [comment]
will look into this tomorrow. it's late here already

/remind me to fix this bug in 18 hourss

**Pull request closed without merging**